panic: failed to apply request nomad.(*nomadFSM).Apply

### Nomad version

`v1.0.4` / `v1.1.2`

### Operating system and Environment details
`Linux ip-<redacted> 4.9.0-16-amd64 #1 SMP Debian 4.9.272-1 (2021-06-21) x86_64 GNU/Linux`

### Issue

While upgrading from v1.0.4. to v1.1.2 I encountered some issues, and lost the cluster leadership, requiring me to perform an [Outage Recovery](https://learn.hashicorp.com/tutorials/nomad/outage-recovery?in=nomad/manage-clusters)

We've got 3 server nodes and were replacing nodes via an aws instance-refresh, going from 3 -> 4, 30 min wait, back to 3,. Something like this
<img width="1032" alt="Screenshot 2021-07-26 at 23 28 59" src="https://user-images.githubusercontent.com/625787/127067280-91c8d808-2cd0-401e-a353-cfa35b57d1ef.png">


When a v1.1.2 node was introduced the v1.0.4 leader crashed with the following logs

```
panic: failed to apply request: []byte{0x2e, 0x84, 0xa6, 0x52, 0x65, 0x67, 0x69, 0x6f, 0x6e, 0xa3, 0x73, 0x74, 0x67, 0xa9, 0x4e, 0x61, 0x6d, 0x65, 0x73, 0x70, 0x61, 0x63, 0x65, 0xa0, 0xa9, 0x41, 0x75, 0x74, 0x68, 0x54, 0x6f, 0x6b, 0x65, 0x6e, 0xda, 0x0, 0x24, 0x30, 0x38, 0x32, 0x62, 0x63, 0x33, 0x35, 0x63, 0x2d, 0x63, 0x64, 0x63, 0x33, 0x2d, 0x65, 0x35, 0x30, 0x39, 0x2d, 0x31, 0x62, 0x65, 0x61, 0x2d, 0x37, 0x34, 0x62, 0x31, 0x63, 0x64, 0x62, 0x38, 0x30, 0x65, 0x65, 0x34, 0xa9, 0x46, 0x6f, 0x72, 0x77, 0x61, 0x72, 0x64, 0x65, 0x64, 0xc2}
--
goroutine 43 [running]:
github.com/hashicorp/nomad/nomad.(*nomadFSM).Apply(0xc0003c2460, 0xc000ff2190, 0x52c4d20, 0xc037da455feffd00)
github.com/hashicorp/nomad/nomad/fsm.go:315 +0x1b17
github.com/hashicorp/raft.(*Raft).runFSM.func1(0xc00036b480)
github.com/hashicorp/raft@v1.1.3-0.20200211192230-365023de17e6/fsm.go:90 +0x2c2
github.com/hashicorp/raft.(*Raft).runFSM.func2(0xc000af5800, 0x40, 0x40)
github.com/hashicorp/raft@v1.1.3-0.20200211192230-365023de17e6/fsm.go:113 +0x75
github.com/hashicorp/raft.(*Raft).runFSM(0xc00105e000)
github.com/hashicorp/raft@v1.1.3-0.20200211192230-365023de17e6/fsm.go:219 +0x3c4
github.com/hashicorp/raft.(*raftState).goFunc.func1(0xc00105e000, 0xc0013f2310)
github.com/hashicorp/raft@v1.1.3-0.20200211192230-365023de17e6/state.go:146 +0x55
created by github.com/hashicorp/raft.(*raftState).goFunc
github.com/hashicorp/raft@v1.1.3-0.20200211192230-365023de17e6/state.go:144 +0x66

```
### Reproduction steps
I haven't attempted to reproduce this yet.

#### Expected Result

No 💥 💣 

#### Actual Result
panic.

### Job file (if appropriate)
n/a

<!--
If possible please post relevant logs in the issue.
`
Logs and other artifacts may also be sent to: nomad-oss-debug@hashicorp.com

Please link to your Github issue in the email and reference it in the subject
line:

> To: nomad-oss-debug@hashicorp.com
>
> Subject: GH-1234: Errors garbage collecting allocs

Emails sent to that address are readable by all HashiCorp employees but are
*not* publicly visible.
-->

### Nomad Server logs (if appropriate)
See above? (I can get more logs if required) 

### Nomad Client logs (if appropriate)
n/a

--- a/.changelog/10952.txt
+++ b/.changelog/10952.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+core: Fixed a panic that may arise when upgrading pre-1.1.0 cluster to 1.1.x and may cause cluster outage
+```

--- a/nomad/acl_endpoint.go
+++ b/nomad/acl_endpoint.go
@@ -851,6 +851,10 @@ func (a *ACL) UpsertOneTimeToken(args *structs.OneTimeTokenUpsertRequest, reply 
 	defer metrics.MeasureSince(
 		[]string{"nomad", "acl", "upsert_one_time_token"}, time.Now())
 
+	if !ServersMeetMinimumVersion(a.srv.Members(), minOneTimeAuthenticationTokenVersion, false) {
+		return fmt.Errorf("All servers should be running version %v or later to use one-time authentication tokens", minAutopilotVersion)
+	}
+
 	// Snapshot the state
 	state, err := a.srv.State().Snapshot()
 	if err != nil {
@@ -898,6 +902,10 @@ func (a *ACL) ExchangeOneTimeToken(args *structs.OneTimeTokenExchangeRequest, re
 	}
 	defer metrics.MeasureSince(
 		[]string{"nomad", "acl", "exchange_one_time_token"}, time.Now())
+
+	if !ServersMeetMinimumVersion(a.srv.Members(), minOneTimeAuthenticationTokenVersion, false) {
+		return fmt.Errorf("All servers should be running version %v or later to use one-time authentication tokens", minAutopilotVersion)
+	}
 
 	// Snapshot the state
 	state, err := a.srv.State().Snapshot()
@@ -951,6 +959,10 @@ func (a *ACL) ExpireOneTimeTokens(args *structs.OneTimeTokenExpireRequest, reply
 	}
 	defer metrics.MeasureSince(
 		[]string{"nomad", "acl", "expire_one_time_tokens"}, time.Now())
+
+	if !ServersMeetMinimumVersion(a.srv.Members(), minOneTimeAuthenticationTokenVersion, false) {
+		return fmt.Errorf("All servers should be running version %v or later to use one-time authentication tokens", minAutopilotVersion)
+	}
 
 	// Check management level permissions
 	if a.srv.config.ACLEnabled {

--- a/nomad/leader.go
+++ b/nomad/leader.go
@@ -48,6 +48,8 @@ var minClusterIDVersion = version.Must(version.NewVersion("0.10.4"))
 
 var minJobRegisterAtomicEvalVersion = version.Must(version.NewVersion("0.12.1"))
 
+var minOneTimeAuthenticationTokenVersion = version.Must(version.NewVersion("1.1.0"))
+
 // monitorLeadership is used to monitor if we acquire or lose our role
 // as the leader in the Raft cluster. There is some work the leader is
 // expected to do, so we must react to changes
@@ -739,6 +741,10 @@ func (s *Server) schedulePeriodic(stopCh chan struct{}) {
 				s.evalBroker.Enqueue(s.coreJobEval(structs.CoreJobCSIVolumeClaimGC, index))
 			}
 		case <-oneTimeTokenGC.C:
+			if !ServersMeetMinimumVersion(s.Members(), minOneTimeAuthenticationTokenVersion, false) {
+				continue
+			}
+
 			if index, ok := getLatest(); ok {
 				s.evalBroker.Enqueue(s.coreJobEval(structs.CoreJobOneTimeTokenGC, index))
 			}

